### PR TITLE
bpftrace: Pull tools into PATH

### PR DESCRIPTION
**Pull request by klemensn**
bpftrace: Pull tools into PATH
    
The *.bt(8) tools are meant to be used as standalone scripts,
their synopsis demonstrates this 

Especially under NixOS, calling them via their absoloute path
in interactive shells is a huge pain, however.
    
Symlink them next to bpftrace(8) itself so they end up in `PATH`
but do not move them to avoid breaking existing setups that expect
them under share/tools/.
    
This goes in line with our bcc as well as Debian's bpftrace package.

@rvl @thoughtpolice @martined

### CODE REVIEW

File: pkgs/os-specific/linux/bpftrace/default.nix
Comment (martinetd) [inline pkgs/os-specific/linux/bpftrace/default.nix:53]: upstream has been reactive with your previous PRs, let's wait until this gets merged imo

File: pkgs/os-specific/linux/bpftrace/default.nix
Comment (martinetd) [inline pkgs/os-specific/linux/bpftrace/default.nix:80]: We don't have anything using these paths in nixpkgs.
Users could depend on it in their repos/scripts but I think it's fine to just move them with a note in next releases notes -- the tools are mostly here for interactive use and not really thought out for scripting in the first place so the impact should be minimal... hopefully? Better than having to keep the links forever anyway.

**Review by martinetd — APPROVED** [review]
debian already installs tools in /usr/sbin so I think it's a good move (ironically now I'm looking they were doing the same rename of man pages to `x.bt.8`...)

btw I also had opened https://github.com/NixOS/nixpkgs/pull/175467 for 0.15.0 update; the commits conflict a bit as I had removed the dump-new patch but new patches should apply fine... Please rebase if mine gets merged first.

I'm also annoyed I didn't noticed ofborg failure in your first PR but now I'm checking it did fail alright :/ this neutral box is misleading...

File: pkgs/os-specific/linux/bpftrace/default.nix
Comment (klemensn) [reply pkgs/os-specific/linux/bpftrace/default.nix:80]: > We don't have anything using these paths in nixpkgs.

If it were just for cases like this, we could obviously move the files and fix all consumers...

> Users could depend on it in their repos/scripts but I think it's fine to just move them with a note in next releases notes -- the tools are mostly here for interactive use and not really thought out for scripting in the first place so the impact should be minimal... hopefully? Better than having to keep the links forever anyway.

This sounds like a viable plan as soon as we sorted out a final solution with upstream.

Until then I'd like to symlink and reduce friction for nixpkgs users (which is too unstable enough already).
This wouldn't be forever.

File: pkgs/os-specific/linux/bpftrace/default.nix
Comment (klemensn) [reply pkgs/os-specific/linux/bpftrace/default.nix:53]: What's the gain?

Given the introduced option stays the same and nothing else gets committed to `doc/CMakeLists.txt` before my PR is merged, fetching `.../pull/2256.pach` would make no difference to `.../commit/<id>.patch` wrt. `sha256`, I believe.

**Comment by klemensn** [comment]
> debian already installs tools in /usr/sbin so I think it's a good move (ironically now I'm looking they were doing the same rename of man pages to `x.bt.8`...)

Good (I didn't look at Debian).

> btw I also had opened #175467 for 0.15.0 update; the commits conflict a bit as I had removed the dump-new patch but new patches should apply fine... Please rebase if mine gets merged first.

Sure.

> I'm also annoyed I didn't noticed ofborg failure in your first PR but now I'm checking it did fail alright :/ this neutral box is misleading...

That failure was not due to my PR, though, was it?

File: pkgs/os-specific/linux/bpftrace/default.nix
Comment (martinetd) [reply pkgs/os-specific/linux/bpftrace/default.nix:53]: on the off chance the commit is refused upstream we'd need to revisite the solution here, it has had no feedback at all yet.

afaik we don't like to diverge from upstream, I personally think these commits don't fix anything urging so would normally say we can wait for 0.15.1 or 0.16 or whatever comes next, but I can agree the release cycle is slow enough so won't push back for merged changes. 

Unmerged changes OTOH should wait, especially since we already have a corresponding rm in the install phase that does exactly the same thing, it's just a cleanup commit not fixing any real problem.

**Comment by martinetd** [comment]
> That failure was not due to my PR, though, was it?

No, and given the nature of your PR we would have merged it anyway (I sure would for this again), but it's still good to notice when tests fails :)

File: pkgs/os-specific/linux/bpftrace/default.nix
Comment (klemensn) [reply pkgs/os-specific/linux/bpftrace/default.nix:53]: > on the off chance the commit is refused upstream we'd need to revisite the solution here, it has had no feedback at all yet.
> 
> afaik we don't like to diverge from upstream, I personally think these commits don't fix anything urging so would normally say we can wait for 0.15.1 or 0.16 or whatever comes next, but I can agree the release cycle is slow enough so won't push back for merged changes.
> 
> Unmerged changes OTOH should wait, especially since we already have a corresponding rm in the install phase that does exactly the same thing, it's just a cleanup commit not fixing any real problem.

Right, I could leave the `fetchpatch`es out and just add this to `postInstall`:
```
ln -s $out/share/bpftrace/tools/*.bt $out/bin/
chmod +x $out/bin/*.bt
```

But I am under the impression that pulling patches is preferred and snippets like the above should always go with comments, which in this case meant links to the PRs/patches I submitted... so my reasoning was:  why not just fetch them directly (documents itself)?

I'm fine with it either way.

**Comment by klemensn** [comment]
> > That failure was not due to my PR, though, was it?
> 
> No, and given the nature of your PR we would have merged it anyway (I sure would for this again), but it's still good to notice when tests fails :)

Good :)

(I did notice a test failure on aarch64 in my last PR but immediately ruled it out and did not investigate^Wdistract myself.)

File: pkgs/os-specific/linux/bpftrace/default.nix
Comment (martinetd) [reply pkgs/os-specific/linux/bpftrace/default.nix:53]: I think it's a bit more subtle than just "preferring patches".

I definitely agree with what's been said in the previous PR: an upstream patch is just easier to maintain. When I or someone does next upgrade we just need to remove the patch and not worry about other modifications to undo (even if in this case there's no real harm in leaving an extra chmod, it's still best not to add it in the first place)

If it's not merged upstream by the time we get it in nixpkgs though it's a lot more work: Next time I upgrade the package I'll have forgotten about this PR (yes, my memory is bad), so I'll need to go look up at either what introduced the fetchpatch or/and the upstream PR to check what is was about, if it was merged without modification or if I need to update something, etc.
As a fetch from PR url it also risks you updating your PR, which will change the hash and make bpftrace unbuildable in further rebuilds.

In this case the whole purpose of this patch is just to replace the `rm -rf $out/share/bpftrace/tools/doc` below which is already here -- just adding a comment it can likely be replaced next update by whatever switch the PR ends up adding is less work for everyone. But maybe I'm just lazy :)

Comment (klemensn) [reply pkgs/os-specific/linux/bpftrace/default.nix]: I just rebased and replaced the patches with extra `postInstall` lines plus comments/links.
This should be simpler and yield less nixpkgs churn.

**Review by martinetd — APPROVED** [review]
Sorry for the delay, I'm still not sure about these links (if there's no deprecation warning it's always going to be a hard call to remove the links later imo -- even if I agree without giving a window for migration it's even worse so it's not a reason not to do it)

Adding a wrapper is probably overkill although it wouldn't be that hard (untested):
```
#!/bin/sh

BINDIR="$(dirname "$0")/../../../bin"
TOOL="$BINDIR/$(basename "$0")"
if ! [ -e "$TOOL" ]; then
    echo "No such tool $TOOL"
    exit 1
fi

echo "Using share/bpftrace/tools/*.bt is deprecated and will be removed, use script in PATH instead" >&2

exec "$TOOL" "$@"
```
with scripts moved to bin, and a link to this (defined as pkgs.writeScript or whatever) for compat?

But, yeah, well, overkill.

Regardless of deprecation warning, this is still better than what we have in all aspects so +1.

**Comment by klemensn** [comment]
> Sorry for the delay, I'm still not sure about these links (if there's no deprecation warning it's always going to be a hard call to remove the links later imo -- even if I agree without giving a window for migration it's even worse so it's not a reason not to do it)
> 
> Adding a wrapper is probably overkill although it wouldn't be that hard (untested):
> 
> ```
> #!/bin/sh
> 
> BINDIR="$(dirname "$0")/../../../bin"
> TOOL="$BINDIR/$(basename "$0")"
> if ! [ -e "$TOOL" ]; then
>     echo "No such tool $TOOL"
>     exit 1
> fi
> 
> echo "Using share/bpftrace/tools/*.bt is deprecated and will be removed, use script in PATH instead" >&2
> 
> exec "$TOOL" "$@"
> ```
> 
> with scripts moved to bin, and a link to this (defined as pkgs.writeScript or whatever) for compat?
> 
> But, yeah, well, overkill.

I don't see any value in such a wrapper.

> Regardless of deprecation warning, this is still better than what we have in all aspects so +1.

Either symlink tools into scope and leave the original scripts in their original paths, causing no harm/churn whatsoever.
Or move scripts into bin, maybe put a line into release notes, and be done with it.

Also, `bptrace` changes most certainly apply to `bcc` as well in nixpkgs: same upstream, same directory/hile structure, but our `bcc` package already symlinks tools and this PR does the same (albeit in a much simpler way) to `bpftrace`.

**Comment by martinetd** [comment]
> I don't see any value in such a wrapper.

providing an alternative without notifying the user about it is useless, we'll have the same breakage you're fearing now when/if we remove the tools directory later, or we'll need to keep the tools directory forever which is bloat.

> Either symlink tools into scope and leave the original scripts in their original paths, causing no harm/churn whatsoever.

The churn is minimal but this is still bloat -- you agreed "This wouldn't be forever."

> Or move scripts into bin, maybe put a line into release notes, and be done with it.

That was my first suggestion :)

> bcc

agreed. I didn't realize bcc already symlinked the tools; I guess we'll want to do the same with it when/if something is decided.

Either way as said at the end of my previous comment this is still an improvement: I'm happy with just a symlink for now, but a cleanup should be discussed/planned/whatever. I don't care if this happens in this PR or an issue or directly in a later PR.

**Comment by klemensn** [comment]
> Either way as said at the end of my previous comment this is still an improvement: I'm happy with just a symlink for now, but a cleanup should be discussed/planned/whatever. I don't care if this happens in this PR or an issue or directly in a later PR.

How about merging this PR now, then we could
- polish bcc's symlinking (looks overly complicated)
- check with upstream about a CMake variable to specify an installation directory
- introduce the same breaking change for bpftrace and bcc in nixpkgs at the same time.

**Comment by martinetd** [comment]
I don't have permissions to merge things, but that was what I meant by it can be done later, and my intention when I gave a "maintainer-approved" review earlier today :)

**Comment by klemensn** [comment]
Anyone willing to merge this?

**Pull request merged**
